<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>How-to work with the rnaseq-pipeline &mdash; rnaseq-pipeline develop documentation</title>
    
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/custom.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     'develop',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="rnaseq-pipeline develop documentation" href="index.html" />
    <link rel="next" title="Setup" href="documentation.html" />
    <link rel="prev" title="Welcome to the documentation of the rnaseq-pipeline" href="index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="documentation.html" title="Setup"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to the documentation of the rnaseq-pipeline"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">rnaseq-pipeline develop documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <p>This document aims to describe how a simple RNAseq analysis can be performed
with the <strong>rnaseq-pipeline</strong>.</p>
<div class="section" id="setup-the-rnaseq-pipeline">
<h1>Setup the <strong>rnaseq-pipeline</strong><a class="headerlink" href="#setup-the-rnaseq-pipeline" title="Permalink to this headline">¶</a></h1>
<div class="section" id="get-the-rnaseq-pipeline">
<h2>Get the <strong>rnaseq-pipeline</strong><a class="headerlink" href="#get-the-rnaseq-pipeline" title="Permalink to this headline">¶</a></h2>
<p>The repository can be obtained like this:</p>
<div class="highlight-python"><pre>$ git clone git@github.com:tiennes/rnaseq-pipeline.git /path/to/repo</pre>
</div>
<p>After cloning the repository, change into the created directory and run the
bootstrapping script to create the required Python environment (which will be
located in <tt class="docutils literal"><span class="pre">/path/to/repo/python_env/</span></tt>):</p>
<div class="highlight-python"><pre>$ cd /path/to/repo
$ ./bootstrap.sh</pre>
</div>
</div>
<div class="section" id="create-your-config-yaml">
<h2>Create your <em>config.yaml</em><a class="headerlink" href="#create-your-config-yaml" title="Permalink to this headline">¶</a></h2>
<p>Start by renaming the file <em>config.sample.yaml</em> into <em>config.yaml</em>:</p>
<div class="highlight-python"><pre>$ cp config.sample.yaml config.yaml</pre>
</div>
<p>The <em>config.yaml</em> file consists of four sections:</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">destination_path</span></tt> &#8211; points to the directory where the result files,
annotations and  temporary files are written to</li>
<li><tt class="docutils literal"><span class="pre">email</span></tt> &#8211; when submitting jobs on a cluster, messages will be sent to
this email address by the cluster scheduler (<a class="reference external" href="mailto:nobody&#37;&#52;&#48;example&#46;com">nobody<span>&#64;</span>example<span>&#46;</span>com</a> by default)</li>
<li><tt class="docutils literal"><span class="pre">tools</span></tt> &#8211; defines all tools used in the pipeline and how to determine
their versions (for later reference)</li>
<li><tt class="docutils literal"><span class="pre">steps</span></tt> &#8211; defines the processing step and their order</li>
</ul>
<p>If you want to know more about the notation that is used in this file, have a
closer look at the <a class="reference external" href="http://www.yaml.org/">YAML definition</a>.</p>
<div class="section" id="sections-of-config-yaml">
<h3>Sections of <strong>config.yaml</strong><a class="headerlink" href="#sections-of-config-yaml" title="Permalink to this headline">¶</a></h3>
<p>huhu</p>
<div class="section" id="section-destination-path">
<h4>Section <em>destination_path</em><a class="headerlink" href="#section-destination-path" title="Permalink to this headline">¶</a></h4>
<p>The value of <tt class="docutils literal"><span class="pre">destination_path</span></tt> is the directory where the pipeline is going
to store the created files. It is possible to use a different directory for
volatile files (see ).</p>
<div class="highlight-yaml"><div class="highlight"><pre><span class="l-Scalar-Plain">destination_path</span><span class="p-Indicator">:</span> <span class="s">&quot;/path/to/dir&quot;</span>
</pre></div>
</div>
</div>
<div class="section" id="section-email">
<h4>Section <em>email</em><a class="headerlink" href="#section-email" title="Permalink to this headline">¶</a></h4>
<p>The value of <tt class="docutils literal"><span class="pre">email</span></tt> is needed if the pipeline is executed on the cluster,
where it is used to inform the person who started the pipeline about status
changes of submitted jobs.</p>
<div class="highlight-yaml"><div class="highlight"><pre><span class="l-Scalar-Plain">email</span><span class="p-Indicator">:</span> <span class="s">&quot;your.name@ufz.de&quot;</span>
</pre></div>
</div>
</div>
<div class="section" id="section-tools">
<h4>Section <em>tools</em><a class="headerlink" href="#section-tools" title="Permalink to this headline">¶</a></h4>
<dl class="docutils">
<dt>The <tt class="docutils literal"><span class="pre">tools</span></tt> block describes all programs needed during the execution of the</dt>
<dd>pipeline.</dd>
</dl>
<div class="highlight-yaml"><div class="highlight"><pre><span class="l-Scalar-Plain">tools</span><span class="p-Indicator">:</span>
    <span class="c1"># you don&#39;t have to specify a path if the tool can be found in $PATH</span>
    <span class="l-Scalar-Plain">cat</span><span class="p-Indicator">:</span>
        <span class="l-Scalar-Plain">path</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">cat</span>
        <span class="l-Scalar-Plain">version</span><span class="p-Indicator">:</span> <span class="s">&quot;--version&quot;</span>
    <span class="c1"># you have to specify a path if the tool can not be found in $PATH</span>
    <span class="l-Scalar-Plain">cutadapt</span><span class="p-Indicator">:</span>
        <span class="l-Scalar-Plain">path</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">/path/to/cutadapt</span>
        <span class="l-Scalar-Plain">version</span><span class="p-Indicator">:</span> <span class="s">&quot;--version&quot;</span>
</pre></div>
</div>
</div>
<div class="section" id="section-steps">
<h4>Section <em>steps</em><a class="headerlink" href="#section-steps" title="Permalink to this headline">¶</a></h4>
<p>The <tt class="docutils literal"><span class="pre">steps</span></tt> block is the core of the pipeline, because it defines the order in
which the different steps of the pipeline are executed. Each step must have a
unique name. Therefore you should give each step a descriptive name followed by
a blank and enclosed in brackets the step type.</p>
<p>There are two different types of steps:</p>
<ol class="arabic">
<li><dl class="first docutils">
<dt><em>sources</em> are used to enter data into the pipeline, meaning they have no</dt>
<dd><p class="first last">predecessor step they depend on.</p>
</dd>
</dl>
</li>
<li><p class="first"><em>steps</em> depend upon a predecessor step and create some output that can be
used by successor steps.</p>
</li>
</ol>
</div>
</div>
<div class="section" id="source-steps">
<h3>Source steps<a class="headerlink" href="#source-steps" title="Permalink to this headline">¶</a></h3>
<p>Source steps provide input files for the other steps within the pipeline, such
as RNA sequences.</p>
<dl class="docutils">
<dt><strong>run_folder_source</strong></dt>
<dd>This source looks for fastq.gz files in
<tt class="docutils literal"><span class="pre">[path]/Unaligned/Project_*/Sample_*</span></tt> and pulls additional information
from CSV sample sheets it finds. It also makes sure that index information
for all samples is coherent and unambiguous.</dd>
<dt><strong>fastq_source</strong></dt>
<dd>This source is similar to the <strong>run_folder_source</strong> but for this source
you have to provide the information which are otherwise contained in the
run folder.</dd>
<dt><strong>raw_file_source</strong></dt>
<dd>This source can provide any file on the filesystem as pipeline input. This
source needs a SHA1 sum of the file</dd>
<dt><strong>raw_url_source</strong></dt>
<dd>This source can provide any downloadable file as pipeline input.</dd>
</dl>
<p>Example configurations for these source steps are shown below:</p>
<div class="highlight-yaml"><div class="highlight"><pre><span class="c1"># sources steps</span>
<span class="l-Scalar-Plain">steps</span><span class="p-Indicator">:</span>
    <span class="c1"># fastq_source provides a number of fastq.gz files as pipeline input</span>
    <span class="l-Scalar-Plain">casava_output (fastq_soure)</span><span class="p-Indicator">:</span>
        <span class="c1"># a glob pattern</span>
        <span class="l-Scalar-Plain">pattern</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">/home/kaempf/Projects/RNAseq_Jurkats+BaP/data/</span>
        <span class="l-Scalar-Plain">group</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">(Sample_COPD_\d+)_R[12]-head.fastq.gz</span>
        <span class="l-Scalar-Plain">indices</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">indices.csv</span>
        <span class="l-Scalar-Plain">paired_end</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">yes</span>

    <span class="c1"># run_folder_source</span>
    <span class="l-Scalar-Plain">fc1 (run_folder_source)</span><span class="p-Indicator">:</span>
        <span class="l-Scalar-Plain">path</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">/data/bioinf/projects/data/Jurkats_BaP_Transcriptome/130108_SN928_0083_AD11VNACXX_Keep/</span>
        <span class="l-Scalar-Plain">paired_end</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">yes</span>
    <span class="l-Scalar-Plain">fc2 (run_folder_source)</span><span class="p-Indicator">:</span>
        <span class="l-Scalar-Plain">path</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">/data/bioinf/projects/data/Jurkats_BaP_Transcriptome/130108_SN928_0084_BC0UT2ACXX_Keep/</span>
        <span class="l-Scalar-Plain">paired_end</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">yes</span>
    <span class="c1"># raw_file_source can provide any filesystem file as pipeline input</span>
    <span class="l-Scalar-Plain">mapped_reads (raw_file_source)</span><span class="p-Indicator">:</span>
        <span class="l-Scalar-Plain">path</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">data/H3K4me3_GCCAAT_L001_001.dup_rm.sam.gz</span>
        <span class="l-Scalar-Plain">sha1</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">835779504aa63f80c9e1008f93f554269d0ec506</span>
    <span class="c1"># raw_url_source can provide any downloadable file as pipeline input</span>
    <span class="l-Scalar-Plain">gencode (raw_url_source)</span><span class="p-Indicator">:</span>
        <span class="l-Scalar-Plain">url</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">ftp://ftp.sanger.ac.uk/pub/gencode/release_15/gencode.v15.annotation.gtf.gz</span>
        <span class="l-Scalar-Plain">sha1</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">9b272fde8bca544e6cd8621ddeec55aa09cf7a05</span>
</pre></div>
</div>
<div class="section" id="run-folder-source">
<h4>Run folder source<a class="headerlink" href="#run-folder-source" title="Permalink to this headline">¶</a></h4>
<p>Here&#8217;s an example:</p>
<div class="highlight-yaml"><div class="highlight"><pre><span class="p-Indicator">-</span> <span class="l-Scalar-Plain">run_folder_source</span><span class="p-Indicator">:</span> <span class="p-Indicator">{</span> <span class="nv">path</span><span class="p-Indicator">:</span> <span class="nv">in</span> <span class="p-Indicator">}</span>
</pre></div>
</div>
<p>This source looks for fastq.gz files in
<tt class="docutils literal"><span class="pre">[path]/Unaligned/Project_*/Sample_*</span></tt> and pulls additional information from
CSV sample sheets it finds.
It also makes sure that index information for all samples is coherent and
unambiguous.</p>
</div>
<div class="section" id="fastq-source">
<h4>FASTQ source<a class="headerlink" href="#fastq-source" title="Permalink to this headline">¶</a></h4>
<p>Here&#8217;s an example:</p>
<div class="highlight-yaml"><div class="highlight"><pre><span class="p-Indicator">-</span> <span class="l-Scalar-Plain">fastq_source</span><span class="p-Indicator">:</span>
    <span class="l-Scalar-Plain">pattern</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">/data/original-fastq/&amp;#42;.fastq.gz</span>
    <span class="l-Scalar-Plain">group</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">(Sample_COPD_\d+)_R[12].fastq.gz</span>
    <span class="l-Scalar-Plain">indices</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">copd-barcodes.csv</span>
</pre></div>
</div>
<p>Input files are collected as defined by <tt class="docutils literal"><span class="pre">pattern</span></tt> and grouped into samples
according to <tt class="docutils literal"><span class="pre">group</span></tt>, which is a regular expression.
All groups defined in the regex <tt class="docutils literal"><span class="pre">(</span>&nbsp; <span class="pre">)</span></tt> are used to construct the sample
name, here it is used to declare that both R1 and R2 files belong to the
same sample.
Indices are read from the CSV file specified by <tt class="docutils literal"><span class="pre">indices</span></tt>.</p>
</div>
</div>
<div class="section" id="normal-steps">
<h3>Normal steps<a class="headerlink" href="#normal-steps" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<div class="section" id="example-configurations">
<h2>Example configurations<a class="headerlink" href="#example-configurations" title="Permalink to this headline">¶</a></h2>
<p>Up until today (June 28th 2013) there is no step to execute Illuminas CASAVA
pipeline, which converts BCL to FASTQ files. Therefore all example configurations
begin with a source step that relies on the availability of fastq.gz files.</p>
<div class="section" id="general-sequencing-analysis-steps">
<h3>General sequencing analysis steps<a class="headerlink" href="#general-sequencing-analysis-steps" title="Permalink to this headline">¶</a></h3>
<p>Every analysis of high-throughput sequencing results starts with some basic
steps. Irrespective of sequencing RNA or DNA, given a reference genome
exists.</p>
<ol class="arabic simple">
<li>Get the sequencing reads as input (most likely fastq.gz)</li>
<li>Remove adapter sequences from your sequencing reads</li>
<li>Align the sequencing reads onto the refernce genome</li>
</ol>
<p>After these steps are finished a lot of different analysis could be applied on
the data. Furtheron example configurations for often used analyses are shown.
The enumeration of steps show continues as if the basic steps were already
performed.</p>
</div>
<div class="section" id="rnaseq-analysis">
<h3>RNAseq analysis<a class="headerlink" href="#rnaseq-analysis" title="Permalink to this headline">¶</a></h3>
<div class="section" id="differential-expression">
<h4>Differential expression<a class="headerlink" href="#differential-expression" title="Permalink to this headline">¶</a></h4>
<p>RNAseq analysis often aims at the discovery of differentially expressed
(known) transcripts. Therefore mappped reads for at least two different samples
have to be available.</p>
<ol class="arabic simple" start="4">
<li>Get annotation set (for e.g. genes, transcripts, ...)</li>
<li>Count the number of reads overlapping the annotation</li>
<li>Perform statistical analysis, based on counts</li>
</ol>
</div>
<div class="section" id="assemble-novel-transcripts">
<h4>Assemble novel transcripts<a class="headerlink" href="#assemble-novel-transcripts" title="Permalink to this headline">¶</a></h4>
<p>As the publicly available annotations, e.g. from GENCODE, are probably not
complete, the assembly of novel transcripts from RNAseq data is another task one
would perform to invetsigate the transcriptome.</p>
</div>
</div>
<div class="section" id="chipseq-analysis">
<h3>ChIPseq analysis<a class="headerlink" href="#chipseq-analysis" title="Permalink to this headline">¶</a></h3>
<p>ChIPseq analysis aims at the discovery of genomic loci at which protein(s) of
interest were bound. The experiment is an enrichment procedure using specific
antibodies. The enrichment detection is normally performed by so called peak
calling programs.</p>
<ol class="arabic simple" start="4">
<li>Get negative control</li>
<li>Peak calling</li>
</ol>
</div>
<div class="section" id="prepare-ucsc-genome-browser-tracks">
<h3>Prepare UCSC genome browser tracks<a class="headerlink" href="#prepare-ucsc-genome-browser-tracks" title="Permalink to this headline">¶</a></h3>
<p>The conversion of sequencing data into an format that can be displayed by the
UCSC genome browser is needed in almost all sequencing projects.</p>
</div>
</div>
</div>
<div class="section" id="monitor-the-rnaseq-pipeline">
<h1>Monitor the <strong>rnaseq-pipeline</strong><a class="headerlink" href="#monitor-the-rnaseq-pipeline" title="Permalink to this headline">¶</a></h1>
</div>
<div class="section" id="scripts">
<h1>Scripts<a class="headerlink" href="#scripts" title="Permalink to this headline">¶</a></h1>
<p>Once the project is set up, there are several scripts which can be used to
execute and monitor the pipeline.
All scripts have a couple of properties in common:</p>
<ul class="simple">
<li>On startup, the configuration is read, tools are checked, input files are
collected, and all tasks are calculated.
If any of these steps fails, the script will print an error message with
a backtrace and it will crash.
This may seem a bit harsh, but after all, it&#8217;s better to fail early than
to fail late if failing is unavoidable.</li>
<li>For convenience, a symbolic link called <tt class="docutils literal"><span class="pre">out</span></tt> will be placed in the
pipeline&#8217;s directory which points to the output directory defined in the
configuration file.
If <tt class="docutils literal"><span class="pre">out</span></tt> already exists, it is left untouched.</li>
</ul>
<p>There are a couple of global command line parameters which are valid for all
scripts (well, actually, it&#8217;s only one):</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal"><span class="pre">--even-if-dirty</span></tt>:</dt>
<dd><p class="first last">Before doing anything else, the pipeline checks whether its source code
has been modified in any way via Git.
If yes, processing is stopped immediately unless this flag is specified.
If you specify the flag, the fact that the repository was dirty will be
recorded in all annotations which are produces <em>including</em> a full Git diff.</p>
</dd>
</dl>
</li>
</ul>
<p>In the following, the scripts are described in detail.</p>
<div class="section" id="status-py">
<h2>status.py<a class="headerlink" href="#status-py" title="Permalink to this headline">¶</a></h2>
<p>The status script lists all tasks resulting from the configured steps and
input samples.
At any time, each task is in one of the following states:</p>
<ul class="simple">
<li><strong>waiting</strong> &#8211; the task is waiting for input files to appear, or its input
files are not up-to-date regarding their respective dependencies</li>
<li><strong>ready</strong> &#8211; all input files are present and up-to-date regarding their
upstream input files (and so on, recursively), the task is ready and can
be started</li>
<li><strong>queued</strong> &#8211; the task is currently queued and will be started &#8220;soon&#8221;
(if you use a computing cluster)</li>
<li><strong>executing</strong> &#8211; the task is currently running on this or another machine</li>
<li><strong>finished</strong> &#8211; all output files are in place and up-to-date</li>
</ul>
<p>Here is an example output:</p>
<div class="highlight-python"><pre>$ ./status.py
Waiting tasks
-------------
[w] cufflinks/Sample_COPD_2023

Ready tasks
-----------
[r] tophat2/Sample_COPD_2023

Finished tasks
--------------
[f] cutadapt/Sample_COPD_2023-R1
[f] cutadapt/Sample_COPD_2023-R2
[f] fix_cutadapt/Sample_COPD_2023

tasks: 5 total, 1 waiting, 1 ready, 3 finished</pre>
</div>
<p>To get a more concise summary, specify <tt class="docutils literal"><span class="pre">--summarize</span></tt>:</p>
<div class="highlight-python"><pre>$ ./status.py --summarize
Waiting tasks
-------------
[w]   1 cufflinks

Ready tasks
-----------
[r]   1 tophat2

Finished tasks
--------------
[f]   2 cutadapt
[f]   1 fix_cutadapt

tasks: 5 total, 1 waiting, 1 ready, 3 finished</pre>
</div>
<p>...or print a fancy ASCII art graph with <tt class="docutils literal"><span class="pre">--graph</span></tt>:</p>
<div class="highlight-python"><pre>$ ./status.py --graph
samples (1 finished)
└─cutadapt (2 finished)
  └─fix_cutadapt (1 finished)
    └─tophat2 (1 ready)
      └─cufflinks (1 waiting)</pre>
</div>
<p>Detailed information about a specific task can be obtained by specifying the
task ID on the command line:</p>
<div class="highlight-python"><pre>$ ./status.py cutadapt/Sample_COPD_2023-R1
info:
  adapter: AGATCGGAAGAGCACACGTCTGAACTCCAGTCACACAGTGATCTCGTATGCCGTCTTCTGCTTG
read_number: R1
output_files:
  log:
    /home/michael/Desktop/rnaseq-pipeline/out/cutadapt-7708/Sample_COPD_2023-cutadapt-R1-log.txt:
    - /home/michael/Desktop/rnaseq-pipeline/copd-small/Sample_COPD_2023_R1.fastq.gz
  reads:
    /home/michael/Desktop/rnaseq-pipeline/out/cutadapt-7708/Sample_COPD_2023-cutadapt-R1.fastq.gz:
    - /home/michael/Desktop/rnaseq-pipeline/copd-small/Sample_COPD_2023_R1.fastq.gz
state: FINISHED</pre>
</div>
<p>This data structure is called the &#8220;run info&#8221; of a certain run and it
represents a kind of plan which includes information about which output
files will be generated and which input files they depend on &#8211; this is
stored in <tt class="docutils literal"><span class="pre">output_files</span></tt>.
Furthermore, necessary information for actually executing the task are
recorded in <tt class="docutils literal"><span class="pre">info</span></tt>.
In this case, the final adapter has been determined by replacing <tt class="docutils literal"><span class="pre">((INDEX))</span></tt>
in the configuration file&#8217;s <tt class="docutils literal"><span class="pre">adapter-R1</span></tt> with the actual barcode index of
the sample.</p>
<p>Because source steps produce no runs and therefore no tasks, they don&#8217;t
appear in the list produced by <tt class="docutils literal"><span class="pre">status.py</span></tt>.
To see their task IDs, specify <tt class="docutils literal"><span class="pre">--sources</span></tt>:</p>
<div class="highlight-python"><pre>$ ./status.py --sources
samples/Sample_COPD_2023</pre>
</div>
<p>You can then specify the ID of a source task like the ID of any other task
to see its details:</p>
<div class="highlight-python"><pre>$ ./status.py samples/Sample_COPD_2023
info:
  index: ACAGTG
  paired_end: true
  read_number:
    Sample_COPD_2023_R1.fastq.gz: R1
    Sample_COPD_2023_R2.fastq.gz: R2
output_files:
  reads:
    /home/michael/Desktop/rnaseq-pipeline/copd-small/Sample_COPD_2023_R1.fastq.gz: []
    /home/michael/Desktop/rnaseq-pipeline/copd-small/Sample_COPD_2023_R2.fastq.gz: []
  state: FINISHED</pre>
</div>
</div>
<div class="section" id="run-locally-py">
<h2>run-locally.py<a class="headerlink" href="#run-locally-py" title="Permalink to this headline">¶</a></h2>
<p>The <tt class="docutils literal"><span class="pre">run-locally.py</span></tt> script runs all non-finished tasks (or a subset)
sequentially on the local machine.
Feel free to cancel this script at any time, it won&#8217;t put your project in a
confused state.
However, if the <tt class="docutils literal"><span class="pre">run-locally.py</span></tt> script receives a SIGKILL signal, the
currently executing job will continue to run and the corresponding task
will be reported as <tt class="docutils literal"><span class="pre">executing</span></tt> by <tt class="docutils literal"><span class="pre">status.py</span></tt> for five more minutes
(SIGTERM should be fine and exit gracefully but <em>doesn&#8217;t just yet</em>).
After that time, you will be warned that a job is marked as being currently
run but no activity has been seen for a while, along with further
instructions about what to do in such a case (don&#8217;t worry, it shouldn&#8217;t
happen by accident).</p>
<p>To execute one or more certain tasks, specify the task IDs on the command
line.
To execute all tasks of a certain step, specify the step name on the command
line.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Why is it safe to cancel the pipeline?
The pipeline is written in a way which expects processes to fail or
cluster jobs to disappear without notice.
This problem is mitigated by a design which relies on file presence and
file timestamps to determine whether a task is finished or not.
Output files are automatically written to temporary locations and later
moved to their real target directory, and it is not until the last file
rename operation has finished that a task is regarded as finished.</p>
</div>
</div>
<div class="section" id="submit-to-cluster-py">
<h2>submit-to-cluster.py<a class="headerlink" href="#submit-to-cluster-py" title="Permalink to this headline">¶</a></h2>
<p>The <tt class="docutils literal"><span class="pre">submit-to-cluster.py</span></tt> script determines which tasks still have to be
carried out and submits the jobs to a GridEngine cluster by calling <tt class="docutils literal"><span class="pre">qsub</span></tt>.
Dependencies are passed to <tt class="docutils literal"><span class="pre">qsub</span></tt> via the <tt class="docutils literal"><span class="pre">-hold_jid</span></tt> option, which means
that jobs that depend on other jobs won&#8217;t get scheduled until their
dependencies have been satisfied.
The file <tt class="docutils literal"><span class="pre">qsub-template.sh</span></tt> is used to submit jobs, with <tt class="docutils literal"><span class="pre">#{</span> <span class="pre">}</span></tt> fields
being substituted with appropriate values.</p>
<p>The file <tt class="docutils literal"><span class="pre">quotas.yaml</span></tt> can be used to define different quotas for different
systems:</p>
<div class="highlight-yaml"><div class="highlight"><pre><span class="s">&quot;frontend[12]&quot;</span><span class="p-Indicator">:</span>
    <span class="l-Scalar-Plain">default</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">5</span>
    <span class="l-Scalar-Plain">cutadapt</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">100</span>
</pre></div>
</div>
<p>In the example above, a default quota of 5 is defined for hosts with a
hostname of <tt class="docutils literal"><span class="pre">frontend1</span></tt> or <tt class="docutils literal"><span class="pre">frontend2</span></tt> (the name is a regular expression).
A quota of 5 means that no more than 5 jobs of one kind will be run in
parallel.
Different quotas can be defined for each step: because <tt class="docutils literal"><span class="pre">cutadapt</span></tt> is
highly I/O-efficient, it has a higher quota.</p>
</div>
</div>
<div class="section" id="annotations">
<h1>Annotations<a class="headerlink" href="#annotations" title="Permalink to this headline">¶</a></h1>
<p>Upon successful completion of a task, an extensive YAML-formatted annotation
is placed next to the output files in a file called
<tt class="docutils literal"><span class="pre">.[task_id]-annotation.yaml</span></tt>.
Also, for every output file, a symbolic link to this file is created:
<tt class="docutils literal"><span class="pre">.[output_filename].annotation.yaml</span></tt>.</p>
<p>Finally, the annotation is rendered via GraphViz, if available.
Rendering can also be done at a later time using annotations as input.
The annotation can be used to determine at a later time what exactly happened.
Also, annotations may help to identify bottlenecks.</p>
<table border="1" class="docutils">
<colgroup>
<col width="45%" />
<col width="55%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="first reference internal image-reference" href="_images/cutadapt.png"><img alt="_images/cutadapt.png" src="_images/cutadapt.png" style="height: 500px;" /></a>
<p class="last">Annotation graph of a <tt class="docutils literal"><span class="pre">cutadapt</span></tt>
run. CPU and RAM usage for individual
processes are shown, file sizes
and line counts are shown for
output files and inter-process
streams.</p>
</td>
<td><a class="first reference internal image-reference" href="_images/cpu-starving.png"><img alt="_images/cpu-starving.png" src="_images/cpu-starving.png" style="height: 500px;" /></a>
<p class="last">In this graph, it becomes evident that
the <tt class="docutils literal"><span class="pre">fix_cutadapt.py</span></tt> process in the middle
gets throttled by the following two <tt class="docutils literal"><span class="pre">pigz</span></tt>
processes, which only run with one core
each and therefore cannot compress the
results fast enough.</p>
</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="extend-the-rnaseq-pipeline">
<h1>Extend the <strong>rnaseq-pipeline</strong><a class="headerlink" href="#extend-the-rnaseq-pipeline" title="Permalink to this headline">¶</a></h1>
<div class="section" id="implement-your-own-steps">
<h2>Implement your own steps<a class="headerlink" href="#implement-your-own-steps" title="Permalink to this headline">¶</a></h2>
<p>The provided pipeline can be easily extended by implementing new steps and
sources. Therefore one does need some basic python programming skills. To add a
new processing step, a single Python file must be placed in <tt class="docutils literal"><span class="pre">include/step</span></tt>
which defines a class with a constructor and two functions. The constructor
(<tt class="docutils literal"><span class="pre">__init__</span></tt>) checks for the availability of required tools and tells the
pipeline which connections this step expects (<tt class="docutils literal"><span class="pre">in/</span></tt>) and which it provides
(<tt class="docutils literal"><span class="pre">out/</span></tt>). The first of the functions  (<tt class="docutils literal"><span class="pre">setup_runs</span></tt>) is used for planning all
jobs based on a list of input files or runs and possibly additional information
from previous steps and the second function (<tt class="docutils literal"><span class="pre">execute</span></tt>) is used to execute a
specific job. The basic scaffold is shown below.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">from</span> <span class="nn">abstract_step</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">import</span> <span class="nn">pipeline</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">process_pool</span>
<span class="kn">import</span> <span class="nn">yaml</span>

<span class="k">class</span> <span class="nc">Macs14</span><span class="p">(</span><span class="n">AbstractStep</span><span class="p">):</span>

    <span class="c"># the constructor</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pipeline</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">Macs14</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">__init__</span><span class="p">(</span><span class="n">pipeline</span><span class="p">)</span>

        <span class="c"># define in and out connections the strings have to start with &#39;in/&#39;</span>
        <span class="c"># or &#39;out/&#39;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">add_connection</span><span class="p">(</span><span class="s">&#39;in/something&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">add_connection</span><span class="p">(</span><span class="s">&#39;out/tag1&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">add_connection</span><span class="p">(</span><span class="s">&#39;out/tag2&#39;</span><span class="p">)</span>
        <span class="o">...</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">require_tool</span><span class="p">(</span><span class="s">&#39;cat4m&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">require_tool</span><span class="p">(</span><span class="s">&#39;pigz&#39;</span><span class="p">)</span>
        <span class="o">...</span>

    <span class="c"># all checks of options and input values should be done here</span>
    <span class="k">def</span> <span class="nf">setup_runs</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">complete_input_run_info</span><span class="p">,</span> <span class="n">connection_info</span><span class="p">):</span>
        <span class="c"># a hash containing information about this step</span>
        <span class="n">output_run_info</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="c"># analyze the complete_input_run_info hash provided by the pipeline</span>
        <span class="k">for</span> <span class="n">step_name</span><span class="p">,</span> <span class="n">step_input_info</span> <span class="ow">in</span> <span class="n">complete_input_run_info</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">input_run_id</span><span class="p">,</span> <span class="n">input_run_info</span> <span class="ow">in</span> <span class="n">step_input_info</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
               <span class="c"># assemble your output_run_info</span>
               <span class="c"># output_run_info has to look like this</span>
               <span class="n">output_run_info</span><span class="p">:</span>
                   <span class="n">run_id_1</span><span class="p">:</span>
                       <span class="s">&quot;output_files&quot;</span><span class="p">:</span>
                           <span class="n">tag1</span><span class="p">:</span>
                               <span class="n">output_file_1</span><span class="p">:</span> <span class="p">[</span><span class="n">input_file_1</span><span class="p">,</span> <span class="n">input_file_2</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span>
                               <span class="n">output_file_2</span><span class="p">:</span> <span class="p">[</span><span class="n">input_file_1</span><span class="p">,</span> <span class="n">input_file_2</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span>
                           <span class="n">tag2</span><span class="p">:</span>
                               <span class="n">output_file_3</span><span class="p">:</span> <span class="p">[</span><span class="n">input_file_1</span><span class="p">,</span> <span class="n">input_file_2</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span>
                               <span class="n">output_file_4</span><span class="p">:</span> <span class="p">[</span><span class="n">input_file_1</span><span class="p">,</span> <span class="n">input_file_2</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span>
                       <span class="s">&quot;info&quot;</span><span class="p">:</span>
                           <span class="o">...</span>
                       <span class="n">more</span><span class="p">:</span>
                           <span class="o">...</span>
                       <span class="n">keys</span><span class="p">:</span>
                           <span class="o">...</span>
                   <span class="n">run_id_2</span><span class="p">:</span>
                       <span class="o">...</span>

        <span class="k">return</span> <span class="n">output_run_info</span>

    <span class="c"># called to actually launch the job (run_info is the hash returned from</span>
    <span class="c"># setup_runs)</span>
    <span class="k">def</span> <span class="nf">execute</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">run_id</span><span class="p">,</span> <span class="n">run_info</span><span class="p">):</span>

        <span class="k">with</span> <span class="n">process_pool</span><span class="o">.</span><span class="n">ProcessPool</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">pool</span><span class="o">.</span><span class="n">Pipeline</span><span class="p">(</span><span class="n">pool</span><span class="p">)</span> <span class="k">as</span> <span class="n">pipeline</span><span class="p">:</span>
                <span class="c"># assemble the steps pipline here</span>
                <span class="n">pipeline</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
                <span class="o">...</span>
                <span class="c"># finally launch it</span>
                <span class="n">pool</span><span class="o">.</span><span class="n">launch</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
</pre></div>
</div>
<p>The code shown above is the framework for a new step. The most essential part is
the hash returned by setup_runs(), here called <tt class="docutils literal"><span class="pre">output_run_info</span></tt>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name"><tt class="docutils literal"><span class="pre">run_id</span></tt>:</th><td class="field-body">It has to be the unique name of a run (obviously, because its a key value).
<tt class="docutils literal"><span class="pre">output_run_info</span></tt> can contain multiple <tt class="docutils literal"><span class="pre">run_id</span></tt> hashes.</td>
</tr>
<tr class="field-even field"><th class="field-name"><tt class="docutils literal"><span class="pre">&quot;output_files&quot;</span></tt>:</th><td class="field-body">This is the only hash key that has to have a fix name. This is used to link
input to output files.</td>
</tr>
<tr class="field-odd field"><th class="field-name"><tt class="docutils literal"><span class="pre">tag[12]</span></tt>:</th><td class="field-body">Every <tt class="docutils literal"><span class="pre">tag</span></tt> has to match <tt class="docutils literal"><span class="pre">\w+$</span></tt> in the string <tt class="docutils literal"><span class="pre">'out/tag'</span></tt>, which was
given to <tt class="docutils literal"><span class="pre">self.add_connection('out/tag')</span></tt>. This can be any string, but it
has to match with the last part of the connection string.</td>
</tr>
<tr class="field-even field"><th class="field-name"><tt class="docutils literal"><span class="pre">output_file_\d</span></tt>:</th><td class="field-body">Each <tt class="docutils literal"><span class="pre">tag</span></tt> has to contain at least one such key. It has to be the name of
the output file produced by the connection <tt class="docutils literal"><span class="pre">'out/tag'</span></tt>. The value of this
has to be a list of related input files. The list can have any number of
entries even zero. Multiple <tt class="docutils literal"><span class="pre">output_file_\d</span></tt> can rely on the same set of
input files.</td>
</tr>
</tbody>
</table>
<p>Also very important is to understand the concept of <em>connections</em>. They provide
input files prior steps created already. The names of the connections can be
arbitrarily chosen, but should <strong>not</strong> describe the file format but more general
terms. For example an <tt class="docutils literal"><span class="pre">out/alignment</span></tt> can provide gzipped SAM or BAM files. So
you have to check in setup runs for the file type provided by a connection and
react accordingly. Inspect <tt class="docutils literal"><span class="pre">complete_input_run_info</span></tt> to find out what your
step gets as input.</p>
<div class="section" id="best-practices">
<h3>Best practices<a class="headerlink" href="#best-practices" title="Permalink to this headline">¶</a></h3>
<p>There are a couple of things which should be kept in mind when implementing new
steps or modifying existing steps:</p>
<ul class="simple">
<li>Make sure errors already show up in <tt class="docutils literal"><span class="pre">setup_runs</span></tt> instead of <tt class="docutils literal"><span class="pre">execute</span></tt>.
Therefore look out for things that may fail in <tt class="docutils literal"><span class="pre">setup_runs</span></tt>. Stick to <em>fail
early, fail often</em>. That way errors show up before submitting jobs to the
cluster and wasting precious cluster waiting time is avoided.</li>
<li>Use the <tt class="docutils literal"><span class="pre">info</span></tt> entry in the returned <tt class="docutils literal"><span class="pre">output_run_info</span></tt> structure to pass
information gathered in <tt class="docutils literal"><span class="pre">setup_runs</span></tt> to <tt class="docutils literal"><span class="pre">execute</span></tt>.</li>
<li>Likewise, make sure that the tools you&#8217;ll need in <tt class="docutils literal"><span class="pre">execute</span></tt> are available.
Check for the availability of tools within the constructor <tt class="docutils literal"><span class="pre">__init__</span></tt>.</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre><span class="c"># make sure tools are available</span>
<span class="bp">self</span><span class="o">.</span><span class="n">require_tool</span><span class="p">(</span><span class="s">&#39;pigz&#39;</span><span class="p">)</span>
<span class="bp">self</span><span class="o">.</span><span class="n">require_tool</span><span class="p">(</span><span class="s">&#39;cutadapt&#39;</span><span class="p">)</span>
</pre></div>
</div>
<ul class="simple">
<li>Make sure your disk access is as cluster-friendly as possible (which
primarily means using large block sizes and preferably no seek operations).
If possible, use <tt class="docutils literal"><span class="pre">unix_pipeline</span></tt> to wrap your commands in <tt class="docutils literal"><span class="pre">pigz</span></tt>, <tt class="docutils literal"><span class="pre">dd</span></tt>,
or <tt class="docutils literal"><span class="pre">cat4m</span></tt> with a large block size like 4 MB.
Although this is not possible in every case (for example when seeking
in files is involved), it is straightforward with tools that read a
continuous stream from <tt class="docutils literal"><span class="pre">stdin</span></tt> and write a continuous stream to
<tt class="docutils literal"><span class="pre">stdout</span></tt>.</li>
</ul>
</div>
</div>
<div class="section" id="add-the-new-step-to-your-configuration">
<h2>Add the new step to your configuration<a class="headerlink" href="#add-the-new-step-to-your-configuration" title="Permalink to this headline">¶</a></h2>
<p>To insert a new step in a pipeline it has to be added into the <tt class="docutils literal"><span class="pre">config.yaml</span></tt>.</p>
</div>
</div>
<div class="section" id="to-do-list">
<h1>To-do list<a class="headerlink" href="#to-do-list" title="Permalink to this headline">¶</a></h1>
<dl class="docutils">
<dt>Timestamps:</dt>
<dd><tt class="docutils literal"><span class="pre">unix_pipeline</span></tt> log messages should include timestamps.</dd>
<dt>Getting started package:</dt>
<dd>We need a small package which demonstrates a quick pipeline, including
the configuration and all required tools.</dd>
<dt>Steps should be able to access all ancestors:</dt>
<dd>All upstream steps should be accessible via their step name or output
file key.</dd>
<dt>On-the-fly steps:</dt>
<dd><p class="first">We need a way to skip writing certain output files and have them flow
temporarily through a pipe only, if possible.
This is a disk space-saving feature only and has no effect on the
outcome of the pipeline. However, it would require that a step is
capable of being run <em>on-the-fly</em> which means it must read and write in
a single stream.</p>
<p>Here&#8217;s an example:</p>
<p class="last graphviz">
<img src="_images/graphviz-663567c7698527d96be3c169697b33db4b725a6b.png" alt="digraph foo {
    rankdir=LR;
    splines=true;
    graph [fontname = Helvetica, fontsize = 12, nodesep = 0.2, ranksep = 0.3];
    node [fontname = Helvetica, fontsize = 12, shape = rect, style=filled, color=&quot;#404040&quot;, fillcolor=&quot;#ffffff&quot;];
    edge [fontname = Helvetica, fontsize = 12, color=&quot;#404040&quot;];

    segemehl [fillcolor = &quot;#fce94f&quot;, color = &quot;#c4a000&quot;];
    in_reads [label = &quot;reads\n(fastq.gz)&quot;];
    mapped_reads [label = &quot;mapped reads\n(sam.gz)&quot;];
    some_filter [fillcolor = &quot;#fce94f&quot;, color = &quot;#c4a000&quot;];
    filtered_reads [label = &quot;filtered reads\n(sam.gz)&quot;];
    htseq_count [label = &quot;htseq-count&quot;, fillcolor = &quot;#fce94f&quot;, color = &quot;#c4a000&quot;];
    counts [label = &quot;counts&quot;];

    in_reads -&gt; segemehl -&gt; mapped_reads -&gt; some_filter -&gt; filtered_reads;
    filtered_reads -&gt; htseq_count -&gt; counts;

    subgraph cluster_food {
        some_filter; filtered_reads;
        label = &quot;on-the-fly step, filtered reads\nnever get written to disk&quot;;
        graph [style=dashed, color=&quot;#808080&quot;];
    }
}" />
</p>
</dd>
<dt>Miscellaneous input files:</dt>
<dd>Genome files and their index such as used by segemehl should not be defined
via a fixed path.
For traceability, it would be preferable to specify the hg19.fa URL and
checksum and have the index generated by a step which the segemehl step
depends on.</dd>
</dl>
<p>Make <tt class="docutils literal"><span class="pre">run-locally.py</span></tt> exit gracefully on receiving SIGTERM.</p>
<dl class="docutils">
<dt>Show statistics for executing tasks:</dt>
<dd>When showing currently executing tasks, show how long this job has already been
running and how it relates to jobs that have already finished.</dd>
</dl>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><a class="reference internal" href="genindex.html"><em>Index</em></a></li>
<li><a class="reference internal" href="py-modindex.html"><em>Module Index</em></a></li>
<li><a class="reference internal" href="search.html"><em>Search Page</em></a></li>
</ul>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Setup the <strong>rnaseq-pipeline</strong></a><ul>
<li><a class="reference internal" href="#get-the-rnaseq-pipeline">Get the <strong>rnaseq-pipeline</strong></a></li>
<li><a class="reference internal" href="#create-your-config-yaml">Create your <em>config.yaml</em></a><ul>
<li><a class="reference internal" href="#sections-of-config-yaml">Sections of <strong>config.yaml</strong></a><ul>
<li><a class="reference internal" href="#section-destination-path">Section <em>destination_path</em></a></li>
<li><a class="reference internal" href="#section-email">Section <em>email</em></a></li>
<li><a class="reference internal" href="#section-tools">Section <em>tools</em></a></li>
<li><a class="reference internal" href="#section-steps">Section <em>steps</em></a></li>
</ul>
</li>
<li><a class="reference internal" href="#source-steps">Source steps</a><ul>
<li><a class="reference internal" href="#run-folder-source">Run folder source</a></li>
<li><a class="reference internal" href="#fastq-source">FASTQ source</a></li>
</ul>
</li>
<li><a class="reference internal" href="#normal-steps">Normal steps</a></li>
</ul>
</li>
<li><a class="reference internal" href="#example-configurations">Example configurations</a><ul>
<li><a class="reference internal" href="#general-sequencing-analysis-steps">General sequencing analysis steps</a></li>
<li><a class="reference internal" href="#rnaseq-analysis">RNAseq analysis</a><ul>
<li><a class="reference internal" href="#differential-expression">Differential expression</a></li>
<li><a class="reference internal" href="#assemble-novel-transcripts">Assemble novel transcripts</a></li>
</ul>
</li>
<li><a class="reference internal" href="#chipseq-analysis">ChIPseq analysis</a></li>
<li><a class="reference internal" href="#prepare-ucsc-genome-browser-tracks">Prepare UCSC genome browser tracks</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#monitor-the-rnaseq-pipeline">Monitor the <strong>rnaseq-pipeline</strong></a></li>
<li><a class="reference internal" href="#scripts">Scripts</a><ul>
<li><a class="reference internal" href="#status-py">status.py</a></li>
<li><a class="reference internal" href="#run-locally-py">run-locally.py</a></li>
<li><a class="reference internal" href="#submit-to-cluster-py">submit-to-cluster.py</a></li>
</ul>
</li>
<li><a class="reference internal" href="#annotations">Annotations</a></li>
<li><a class="reference internal" href="#extend-the-rnaseq-pipeline">Extend the <strong>rnaseq-pipeline</strong></a><ul>
<li><a class="reference internal" href="#implement-your-own-steps">Implement your own steps</a><ul>
<li><a class="reference internal" href="#best-practices">Best practices</a></li>
</ul>
</li>
<li><a class="reference internal" href="#add-the-new-step-to-your-configuration">Add the new step to your configuration</a></li>
</ul>
</li>
<li><a class="reference internal" href="#to-do-list">To-do list</a></li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">Welcome to the documentation of the <strong>rnaseq-pipeline</strong></a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="documentation.html"
                        title="next chapter">Setup</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/how-to.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="documentation.html" title="Setup"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to the documentation of the rnaseq-pipeline"
             >previous</a> |</li>
        <li><a href="index.html">rnaseq-pipeline develop documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, Michael Specht.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2b1.
    </div>
  </body>
</html>