<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>rnaseq-pipeline &mdash; rnaseq-pipeline develop documentation</title>
    
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/custom.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     'develop',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="rnaseq-pipeline develop documentation" href="#" />
    <link rel="next" title="Setup the rnaseq-pipeline" href="how-to.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="how-to.html" title="Setup the rnaseq-pipeline"
             accesskey="N">next</a> |</li>
        <li><a href="#">rnaseq-pipeline develop documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="welcome-to-the-documentation-of-the-rnaseq-pipeline">
<h1>Welcome to the documentation of the <strong>rnaseq-pipeline</strong><a class="headerlink" href="#welcome-to-the-documentation-of-the-rnaseq-pipeline" title="Permalink to this headline">¶</a></h1>
<p>The <strong>rnaseq-pipeline</strong> is a tool to configure, run, and control your high-throughput
sequencing analysis.</p>
<div class="section" id="contents">
<h2>Contents<a class="headerlink" href="#contents" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="how-to.html">Setup the <strong>rnaseq-pipeline</strong></a><ul>
<li class="toctree-l2"><a class="reference internal" href="how-to.html#get-the-rnaseq-pipeline">Get the <strong>rnaseq-pipeline</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="how-to.html#create-your-config-yaml">Create your <em>config.yaml</em></a></li>
<li class="toctree-l2"><a class="reference internal" href="how-to.html#example-configurations">Example configurations</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="how-to.html#monitor-the-rnaseq-pipeline">Monitor the <strong>rnaseq-pipeline</strong></a></li>
<li class="toctree-l1"><a class="reference internal" href="how-to.html#scripts">Scripts</a><ul>
<li class="toctree-l2"><a class="reference internal" href="how-to.html#status-py">status.py</a></li>
<li class="toctree-l2"><a class="reference internal" href="how-to.html#run-locally-py">run-locally.py</a></li>
<li class="toctree-l2"><a class="reference internal" href="how-to.html#submit-to-cluster-py">submit-to-cluster.py</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="how-to.html#annotations">Annotations</a></li>
<li class="toctree-l1"><a class="reference internal" href="how-to.html#extend-the-rnaseq-pipeline">Extend the <strong>rnaseq-pipeline</strong></a><ul>
<li class="toctree-l2"><a class="reference internal" href="how-to.html#implement-your-own-steps">Implement your own steps</a></li>
<li class="toctree-l2"><a class="reference internal" href="how-to.html#add-the-new-step-to-your-configuration">Add the new step to your configuration</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="how-to.html#to-do-list">To-do list</a></li>
<li class="toctree-l1"><a class="reference internal" href="how-to.html#indices-and-tables">Indices and tables</a></li>
<li class="toctree-l1"><a class="reference internal" href="documentation.html">Setup</a><ul>
<li class="toctree-l2"><a class="reference internal" href="documentation.html#the-configuration-file">The configuration file</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="documentation.html#indices-and-tables">Indices and tables</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API documentation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="api.html#pipeline-specific-modules">Pipeline-specific modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="api.html#miscellaneous-modules">Miscellaneous modules</a></li>
</ul>
</li>
</ul>
</div>
</div>
</div>
<div class="section" id="introduction">
<h1>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h1>
<p>The aim of this data processing pipeline is to enable robust and straightforward
bioinformatics data evaluation. It is implemented in Python, runs under
GNU/Linux and can be controlled from the command-line interface. Although the
primary focus is the evaluation of RNASeq data, its design allows for a variety
of other applications.</p>
<div class="section" id="general-usage">
<h2>General usage<a class="headerlink" href="#general-usage" title="Permalink to this headline">¶</a></h2>
<p>This package <em>does not</em> provide a number of tools which are downloaded and
installed system-wide to provide certain functioniality.
The intention of this system is to provide a robust and traceable framework
for data evaluation in scientific experiments which uses other tools and
manages individual data processing steps and their inter-dependencies.</p>
<p>The recommended workflow for running a data evaluation for an experiment is
as follows:</p>
<ol class="arabic simple">
<li>Check-out the rnaseq-pipeline repository via Git.</li>
<li>Setup the project by writing the configuration file.</li>
<li>Add steps or other functionality as needed (optional).</li>
<li>Run the pipeline.</li>
<li>Have your changes (if there are any) merged back into the main repository,
to the advantage of the scientific community.</li>
</ol>
<p>This leaves you with:</p>
<ul class="simple">
<li>Your original input files, which are left untouched.</li>
<li>The experiment-specific pipeline repository.
You should keep this repository for later reference and you could even
make it publicly available along with your input files for anybody to
re-run the entire data evaluation or parts thereof.</li>
<li>The output directory containing all output files and comprehensive
annotations.
These annotations include detailed information for every output file,
including which steps have been executed and the Git SHA1 hash of
the pipeline repository at the time the data processing took place.
In many cases, these annotations also include information about all
inter-process streams and output files, including SHA1 checksums, file
sizes, and line counts.</li>
</ul>
</div>
<div class="section" id="core-aspects">
<h2>Core aspects<a class="headerlink" href="#core-aspects" title="Permalink to this headline">¶</a></h2>
<p><strong>Robustness:</strong></p>
<ul class="simple">
<li>All steps write their output files to a temporary location.
Only if a step has completed successfully, the output files are copied to
the correct output directory.</li>
<li>The output directory names are suffixed with a four-character hashtag
which mirrors the options specified for the step.</li>
<li>Processing can be aborted and continued from the command line at any time.
This way, cluster failures are less critical because output files do not
get compromised.</li>
<li>Errors are caught as early as possible. Tools are checked for availability,
and the entire processing pipeline is calculated in advance before
jobs are being started or submitted to a cluster.</li>
</ul>
<p><strong>Traceability:</strong></p>
<ul class="simple">
<li>Comprehensive annotations are written to the output directories, allowing
for later investigation about what exactly happened.</li>
</ul>
<p><strong>Simplicity:</strong></p>
<ul class="simple">
<li>The entire processing pipeline is described via a configuration file.
Steps are defined in a directed acyclic graph (DAG).</li>
<li>Interaction with the pipeline happens through a handful of scripts which
are used to monitor the state of the pipeline and execute individual or all
remaining steps.</li>
</ul>
</div>
<div class="section" id="design">
<h2>Design<a class="headerlink" href="#design" title="Permalink to this headline">¶</a></h2>
<p>The central part of the pipeline is its definition of the steps which are to
be carried out.
Steps are organized in a dependency graph (a directed acyclic graph) &#8211; every
step may have one or more parent steps, which may in turn have other parent
steps, and so on.
Steps without parents are usually sources which provide source files, for
example FASTQ files with the raw sequences obtained from the sequencer,
genome sequence databases or annotation tracks.</p>
<p>Each step defines a number of runs and each run represents a piece of the
entire data evaluation, typically at the level of a single sample.
A certain <em>run</em> of a certain <em>step</em> is called a <em>task</em>.
While the steps only describe what needs to be done on a very abstract level,
it is through the individual runs of each step that a pipeline-wide list of
actual tasks becomes available.
Each run may provide a number of output files which depend on output files
of one or several runs from parent steps.</p>
<p>Source steps define a run for every input sample, and a subsequent step
may:</p>
<ul class="simple">
<li>define the same number of runs,</li>
<li>define more runs (for example when R1 and R2 reads in a paired-end RNASeq
experiment should be treated separately),</li>
<li>define fewer runs (usually towards the end of a pipeline, where results are
summarized).</li>
</ul>
</div>
</div>
<div class="section" id="remarks">
<h1>Remarks<a class="headerlink" href="#remarks" title="Permalink to this headline">¶</a></h1>
<p>This documentation has been created using <a class="reference external" href="http://sphinx-doc.org/">Sphinx</a>
and <a class="reference external" href="http://docutils.sourceforge.net/rst.html">reStructuredText</a>.</p>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><a class="reference internal" href="genindex.html"><em>Index</em></a></li>
<li><a class="reference internal" href="py-modindex.html"><em>Module Index</em></a></li>
<li><a class="reference internal" href="search.html"><em>Search Page</em></a></li>
</ul>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Welcome to the documentation of the <strong>rnaseq-pipeline</strong></a><ul>
<li><a class="reference internal" href="#contents">Contents</a><ul>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#introduction">Introduction</a><ul>
<li><a class="reference internal" href="#general-usage">General usage</a></li>
<li><a class="reference internal" href="#core-aspects">Core aspects</a></li>
<li><a class="reference internal" href="#design">Design</a></li>
</ul>
</li>
<li><a class="reference internal" href="#remarks">Remarks</a></li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>

  <h4>Next topic</h4>
  <p class="topless"><a href="how-to.html"
                        title="next chapter">Setup the <strong>rnaseq-pipeline</strong></a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/index.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="how-to.html" title="Setup the rnaseq-pipeline"
             >next</a> |</li>
        <li><a href="#">rnaseq-pipeline develop documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, Michael Specht.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2b1.
    </div>
  </body>
</html>